Sources/Network/NetworkData.swift

```diff
@@ -59,6 +72,32 @@ open class NetworkDataProcedure<Session: URLSessionTaskFactory>: Procedure, Resu
             strongSelf.finish()
         }
 
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToSend), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesSent), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToReceive), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesReceived), options: [.new, .old], context: &observerContext)
+
         task?.resume()
     }
+
+    override open func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        guard context == &observerContext else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+            return
+        }
+
+        if let keyPath = keyPath, let task = object as? Session.DataTask {
+            switch keyPath {
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToSend):
+                progress.totalUnitCount = task.countOfBytesExpectedToSend
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesSent):
+                progress.completedUnitCount = task.countOfBytesSent
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToReceive):
+                progress.totalUnitCount = task.countOfBytesExpectedToReceive
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesReceived):
+                progress.completedUnitCount = task.countOfBytesReceived
+            default: break
```

Comment:
According to the docs for [NSURLSessionDataTask](https://developer.apple.com/reference/foundation/nsurlsessiondatatask), it would seem as though a data task moves through the following flow:

- During upload of the body data (if your app provides any), the session periodically calls its delegate’s `URLSession:task:didSendBodyData:totalBytesSent:totalBytesExpectedToSend:` method with status information.
**NOTE: According to the docs for [countOfBytesSent](https://developer.apple.com/reference/foundation/urlsessiontask/1410444-countofbytessent), this should map to `countOfBytesSent` being incremented.**

- After receiving an initial response, the session calls its delegate’s `URLSession:dataTask:didReceiveResponse:completionHandler:` method to let you examine the status code and headers, and optionally convert the data task into a download task.
**NOTE: According to the docs for [countOfBytesExpectedToReceive](https://developer.apple.com/reference/foundation/urlsessiontask/1410663-countofbytesexpectedtoreceive), this should map to `countOfBytesExpectedToReceive` being updated to whatever the `Content-Length` header of the server's response indicates.**

- During the transfer [of the response], the session calls its delegate’s `URLSession:dataTask:didReceiveData:` method to provide your app with the content as it arrives.
**NOTE: According to the docs for [countOfBytesReceived](https://developer.apple.com/reference/foundation/urlsessiontask/1411581-countofbytesreceived), this should map to `countOfBytesReceived` being incremented.**